universe/Pathfinder.cpp

```diff
@@ -907,6 +1012,147 @@ std::multimap<double, int> Pathfinder::PathfinderImpl::ImmediateNeighbors(int sy
     return std::multimap<double, int>();
 }
 
+
+/** Examine a single universe object and determine if it is within jumps
+    of any object in others.*/
+struct WithinJumpsOfOthersObjectVisitor : public boost::static_visitor<bool> {
+    WithinJumpsOfOthersObjectVisitor(const Pathfinder::PathfinderImpl& _pf,
+                                     int _jumps,
+                                     const std::vector<std::shared_ptr<const UniverseObject>>& _others) :
+        pf(_pf), jumps(_jumps), others(_others)
+    {}
+
+    bool operator()(std::nullptr_t) const { return false; }
+    bool operator()(int sys_id) const {
+        bool retval = pf.WithinJumpsOfOthers(jumps, sys_id, others);
+        return retval;
+    }
+    bool operator()(std::pair<int, int> prev_next) const {
+        return pf.WithinJumpsOfOthers(jumps, prev_next.first, others)
+            || pf.WithinJumpsOfOthers(jumps, prev_next.second, others);
+    }
+    const Pathfinder::PathfinderImpl& pf;
+    int jumps;
+    const std::vector<std::shared_ptr<const UniverseObject>>& others;
+};
+
+/** Examine a single other in the cache to see if any of its locations
+    are within jumps*/
+struct WithinJumpsOfOthersOtherVisitor : public boost::static_visitor<bool> {
+    WithinJumpsOfOthersOtherVisitor(const Pathfinder::PathfinderImpl& _pf,
+                                    int _jumps,
+                                    const distance_matrix_storage<short>::row_ref _row) :
+        pf(_pf), jumps(_jumps), row(_row)
+    {}
+
+    bool single_result(int other_id) const {
+        int index;
+        try {
+            index = pf.m_system_id_to_graph_index.at(other_id);
+        } catch (const std::out_of_range& e) {
+            ErrorLogger() << "Passed invalid system id: " << other_id;
+            return false;
+        }
+        bool retval = (row[index] <= jumps);
+        return retval;
+    }
+
+    bool operator()(std::nullptr_t) const { return false; }
+    bool operator()(int sys_id) const {
+        return single_result(sys_id);
+    }
+    bool operator()(std::pair<int, int> prev_next) const {
+        return single_result(prev_next.first)
+            || single_result(prev_next.second);
+    }
+    const Pathfinder::PathfinderImpl& pf;
+    int jumps;
+    const distance_matrix_storage<short>::row_ref row;
+};
+
+
+void Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit(
+    bool& answer,
+    int jumps,
+    const std::vector<std::shared_ptr<const UniverseObject>>& others,
+    size_t ii, const distance_matrix_storage<short>::row_ref row) const
```

Comment:
I changed it to
````
void Pathfinder::PathfinderImpl::WithinJumpsOfOthersCacheHit(
    bool& answer, int jumps,
    const std::vector<std::shared_ptr<const UniverseObject>>& others,
    size_t ii, const distance_matrix_storage<short>::row_ref row) const
````
I was trying to keep that last two parameters on the same line.  Those are the two parameters required for the cache hit handler signature.  All the other parameters need to be bound to something before the cache hit handler is passed to the `distance_matrix_cache`.